generator/src/main/java/com/graphicsfuzz/generator/mutateapi/ExpressionGenerator.java

```diff
@@ -0,0 +1,629 @@
+import com.graphicsfuzz.generator.semanticschanging.LiteralFuzzer;
+import com.graphicsfuzz.util.Constants;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import org.apache.commons.lang3.tuple.ImmutablePair;
+import org.apache.commons.lang3.tuple.Pair;
+
+public class ExpressionGenerator {
+
+  private static final int MAX_FUNCTION_PARAMS = 5;
+  private static final int NUM_WAYS_TO_GENERATE_EXPR = 7;
+  private static final int MAX_DEPTH = 5;
+  // Theses boundaries are taken from the LiteralFuzzer, we may want to consider
+  // modifying these values.
+  private static final int INT_MIN = 0;
+  private static final int INT_MAX = 1 << 17;
+
+  private static final String NEGATIVE = "_NEGATIVE";
+  private static final String ID = "_id";
+
+  private final TranslationUnit translationUnit;
+  private final PipelineInfo pipelineInfo;
+  private final IdGenerator idGenerator;
+  private final IRandom generator;
+  private int currentDepth;
+  private final LiteralFuzzer literalFuzzer;
+  private final FactManager globalFactManager;
+
+
+  public ExpressionGenerator(TranslationUnit translationUnit, PipelineInfo pipelineInfo,
+                             IRandom generator,
+                             FactManager globalFactManager) {
+    this.translationUnit = translationUnit;
+    this.pipelineInfo = pipelineInfo;
+    this.idGenerator = new IdGenerator();
+    this.generator = generator;
+    this.currentDepth = 0;
+    this.literalFuzzer = new LiteralFuzzer(this.generator);
+    this.globalFactManager = globalFactManager;
+  }
+
+  private Expr generateLiteralValue(Value value) {
+    return value.generateLiteral(literalFuzzer);
+  }
+
+  /**
+   * Given the expected value, this function will recursively declare variable declarations or
+   * generate new functions whose results are known by the fact manager. If the given value is
+   * empty, the fact manager can generate any arbitrary values that have the correct type.
+   *
+   * @param factManager        manager class holding the value and its associated expression that
+   *                           guarantees to compute the given value.
+   * @param functionDefinition a function into which the new expression will be injected.
+   * @param stmt               statement in the body of the given function where the generated
+   *                           expression will be inserted before.
+   * @param value              the value that will be computed by the expression generated by
+   *                           this method.
+   * @return the expression whose value is known by the fact manager.
+   */
+  public Expr generateExpr(FactManager factManager,
+                           FunctionDefinition functionDefinition,
+                           Stmt stmt,
+                           Value value) {
+
+    if (currentDepth > MAX_DEPTH) {
```

Comment:
"the maximum limitation" -> "the maximum limit"

"we don't want the program to go deeper" -> "we don't want generation to go deeper"